File: backend_python/agents/risk_grader.py
"""
KawalKontrak.ai — Agent 3: Risk Grader
========================================

Tanggung Jawab:
    Mengevaluasi risiko dari setiap klausul yang sudah dicocokkan
    dengan hukum. Agen ini bertugas:
    1. Membuat red flag dengan penjelasan bahasa awam dan analogi sederhana.
    2. Mengidentifikasi klausul yang sudah aman (safe clauses).
    3. Memberikan estimasi awal risk_level (akan di-override deterministik
       oleh orchestrator berdasarkan severity tertinggi — lihat E1b).
    4. Menyusun ringkasan kontrak dan langkah aksi untuk pekerja.

    Catatan: klausul berstatus TIDAK_DITEMUKAN TIDAK diproses di sini —
    orchestrator memasukkannya ke daftar 'klausul_tinjauan' agar tetap
    terlihat pengguna (FR-05), bukan hilang diam-diam.

Model:  KK_MODEL_CORE (default gemini-2.5-flash — reasoning mendalam)
Input:  list[MatchedClause]
Output: RiskGraderOutput
"""

import json
import logging
from tenacity import retry, stop_after_attempt, wait_exponential

from google import genai
from google.genai import types

from backend_python.config import MODEL_CORE
from backend_python.models import MatchedClause, RiskGraderOutput
from backend_python.utils import locale_instruction

logger = logging.getLogger(__name__)

_SYSTEM_PROMPT = """
Anda adalah Asisten Hukum Ketenagakerjaan yang empatik dan komunikatif.
Target pembaca Anda adalah pekerja awam yang tidak berlatar belakang hukum.

Tugas Anda:
1. Untuk setiap klausul berstatus MELANGGAR atau AMBIGU:
   - Buat 'red flag' dengan bahasa yang jelas dan mudah dipahami.
   - Sertakan analogi kehidupan sehari-hari yang relatable.
   - Beri rekomendasi negosiasi yang praktis dan realistis.
2. Untuk setiap klausul berstatus SESUAI:
   - Buat 'klausul aman' dengan penjelasan singkat kenapa ini baik.
3. Abaikan klausul berstatus TIDAK_DITEMUKAN — klausul tersebut sudah
   ditangani terpisah sebagai "perlu tinjauan manusia".
4. Tentukan level risiko keseluruhan kontrak (estimasi awal — sistem akan menghitung ulang secara deterministik).
5. Susun ringkasan kontrak dan daftar langkah aksi berurutan untuk pekerja.

RUBRIK SEVERITY — WAJIB DIIKUTI (menentukan tingkat keparahan tiap red flag):

| status_hukum | Kondisi Klausul                                                                 | Severity yang Ditetapkan          |
|---|---|---|
| MELANGGAR    | Menyangkut hak dasar: upah/gaji, pesangon, kompensasi PKWT, PHK sepihak        | CRITICAL                          |
| MELANGGAR    | Menyangkut hal prosedural/administratif (format, bahasa kontrak, dll.)          | HIGH atau MEDIUM                  |
| AMBIGU       | Berpotensi merugikan tapi membutuhkan konteks tambahan                          | MEDIUM (DILARANG CRITICAL — status belum pasti melanggar) |
| AMBIGU       | Klausul wajib (kompensasi, pengupahan, jam kerja) tidak disebutkan sama sekali  | HIGH atau MEDIUM                  |

ATURAN KONSISTENSI SEVERITY (WAJIB):
- Klausul MELANGGAR tidak boleh diberi severity lebih rendah dari klausul AMBIGU sejenis.
- Klausul AMBIGU DILARANG diberi severity CRITICAL — status belum pasti melanggar.
- Jika ragu antara dua severity yang berdekatan, pilih yang lebih tinggi untuk klausul yang memengaruhi upah/status kerja pekerja.

Prinsip penulisan:
- Gunakan kata 'Anda' saat merujuk pekerja, bukan 'kamu' atau 'karyawan'.
- Hindari jargon hukum — jelaskan dalam kalimat sederhana.
- Tone: profesional namun hangat, waspada namun tidak menakut-nakuti.
""".strip()


@retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=2, max=10), reraise=True)
async def grade_risks(
    matched_clauses: list[MatchedClause],
    client: genai.Client,
    locale: str = "en",
) -> RiskGraderOutput | None:
    """
    Menilai risiko seluruh klausul dan menyusun laporan analisis lengkap (async).

    Args:
        matched_clauses: Daftar klausul + status hukum dari Legal Matcher.
        client:          Instance Google GenAI client.
        locale:          Bahasa nilai output.

    Returns:
        RiskGraderOutput, atau None jika terjadi error fatal.
        Catatan: field `risk_level` di output ini adalah estimasi model —
        orchestrator akan menggantinya dengan nilai deterministik (E1b fix).
    """
    if not matched_clauses:
        logger.warning("Risk Grader: Daftar klausul kosong, tidak ada yang dinilai.")
        return None

    logger.info(f"Risk Grader: Menilai {len(matched_clauses)} klausul...")

    # Serialisasi matched_clauses ke JSON yang mudah dibaca model
    payload = json.dumps(
        [c.model_dump() for c in matched_clauses],
        ensure_ascii=False,
        indent=2,
    )

    try:
        response = await client.aio.models.generate_content(
            model=MODEL_CORE,
            contents=(
                f"Berikut adalah daftar klausul kontrak beserta status hukum "
                f"dan referensi peraturannya. Evaluasi dan buat laporan risiko:\n\n"
                f"{payload}"
            ),
            config=types.GenerateContentConfig(
                system_instruction=_SYSTEM_PROMPT + locale_instruction(locale),
                temperature=0.3,  # Sedikit lebih tinggi agar penjelasan lebih natural
                response_mime_type="application/json",
                response_schema=RiskGraderOutput,
            ),
        )

        result: RiskGraderOutput = response.parsed  # type: ignore[assignment]

        logger.info(
            f"Risk Grader: Selesai — {len(result.red_flags)} red flags, "
            f"{len(result.klausul_aman)} klausul aman, "
            f"level risiko model: {result.risk_level.value} "
            f"(akan di-override deterministik oleh orchestrator)."
        )
        return result

    except Exception as exc:
        logger.error(f"Risk Grader: Gagal — {exc}", exc_info=True)
        return None

File: src/lib/agents/risk-grader.ts
import { GoogleGenAI, Type } from '@google/genai';
import { MatchedClause } from './legal-matcher';
import { RedFlag, SafeClause, ContractSummary, RiskLevel } from '@/types';

export interface RiskGraderOutput {
  red_flags: Omit<RedFlag, 'email_template'>[];
  klausul_aman: SafeClause[];
  ringkasan: ContractSummary;
  risk_level: RiskLevel;
  langkah_berikutnya: string[];
}

export async function gradeRisks(
  matchedClauses: MatchedClause[], 
  ai: GoogleGenAI
): Promise<RiskGraderOutput | null> {
  
  const payload = JSON.stringify(matchedClauses, null, 2);
  
  const response = await ai.models.generateContent({
    model: 'gemini-2.5-flash',
    contents: `Anda adalah Asisten Hukum Ketenagakerjaan. Berikut adalah daftar klausul kontrak beserta status hukum dan referensi peraturannya. 
Tugas Anda adalah mengevaluasi risiko dari keseluruhan kontrak ini dan menyusun laporannya.
Untuk setiap klausul yang MELANGGAR hukum, buatlah RedFlag dengan bahasa awam yang mudah dipahami, sertakan analogi sederhana.
Untuk setiap klausul yang SESUAI, buatlah SafeClause.
Berikan juga ringkasan jenis kontrak, level risiko keseluruhan, dan langkah selanjutnya yang harus dilakukan pekerja.

Data Klausul:
${payload}`,
    config: {
      temperature: 0.2,
      responseMimeType: 'application/json',
      responseSchema: {
        type: Type.OBJECT,
        properties: {
          red_flags: {
            type: Type.ARRAY,
            items: {
              type: Type.OBJECT,
              properties: {
                flag_id: { type: Type.STRING },
                severity: { type: Type.STRING, enum: ["CRITICAL", "HIGH", "MEDIUM", "LOW"] },
                pasal_kontrak: { type: Type.STRING },
                potensi_masalah: { type: Type.STRING, description: "Penjelasan bahasa awam" },
                referensi_uu: {
                  type: Type.ARRAY,
                  items: {
                    type: Type.OBJECT,
                    properties: {
                      peraturan: { type: Type.STRING },
                      pasal: { type: Type.STRING },
                      judul: { type: Type.STRING },
                      ketentuan_relevan: { type: Type.STRING }
                    },
                    required: ["peraturan", "pasal", "judul", "ketentuan_relevan"]
                  }
                },
                rekomendasi_negosiasi: { type: Type.STRING },
                analogi_sederhana: { type: Type.STRING }
              },
              required: ["flag_id", "severity", "pasal_kontrak", "potensi_masalah", "referensi_uu", "rekomendasi_negosiasi", "analogi_sederhana"]
            }
          },
          klausul_aman: {
            type: Type.ARRAY,
            items: {
              type: Type.OBJECT,
              properties: {
                pasal_kontrak: { type: Type.STRING },
                terjemahan: { type: Type.STRING },
                referensi: {
                  type: Type.ARRAY,
                  items: {
                    type: Type.OBJECT,
                    properties: {
                      peraturan: { type: Type.STRING },
                      pasal: { type: Type.STRING },
                      judul: { type: Type.STRING },
                      ketentuan_relevan: { type: Type.STRING }
                    },
                    required: ["peraturan", "pasal", "judul", "ketentuan_relevan"]
                  }
                }
              },
              required: ["pasal_kontrak", "terjemahan"]
            }
          },
          ringkasan: {
            type: Type.OBJECT,
            properties: {
              jenis: { type: Type.STRING },
              durasi: { type: Type.STRING },
              gaji_bulanan: { type: Type.STRING },
              status: { type: Type.STRING },
              harus_diubah: { type: Type.ARRAY, items: { type: Type.STRING } },
              sebaiknya_diubah: { type: Type.ARRAY, items: { type: Type.STRING } }
            },
            required: ["jenis", "status", "harus_diubah", "sebaiknya_diubah"]
          },
          risk_level: { type: Type.STRING, enum: ["CRITICAL", "HIGH", "MEDIUM", "LOW"] },
          langkah_berikutnya: { type: Type.ARRAY, items: { type: Type.STRING } }
        },
        required: ["red_flags", "klausul_aman", "ringkasan", "risk_level", "langkah_berikutnya"]
      }
    }
  });

  if (!response.text) return null;
  
  try {
    return JSON.parse(response.text) as RiskGraderOutput;
  } catch (e) {
    console.error("Gagal parse Risk-Grader output", e);
    return null;
  }
}
